File: server/controllers/authController.js
import crypto from 'crypto';
import asyncHandler from 'express-async-handler';
import { validationResult } from 'express-validator';
import User from '../models/User.js';
import generateToken from '../utils/generateToken.js';
import { createNotification } from '../utils/notificationHelper.js';
import sendEmail from '../utils/sendEmail.js';

const OTP_TTL_MINUTES = Number(process.env.OTP_EXPIRES_MINUTES) || 10;

const setCookieAndRespond = (res, statusCode, user, token) => {
  const cookieOptions = {
    httpOnly: true,
    secure: process.env.NODE_ENV === 'production',
    sameSite: 'strict',
    maxAge: 7 * 24 * 60 * 60 * 1000, // 7 days
  };
  res.cookie('token', token, cookieOptions);
  res.status(statusCode).json({
    success: true,
    token,
    user: {
      id: user._id,
      name: user.name,
      email: user.email,
      role: user.role,
      ownerRequestStatus: user.ownerRequestStatus,
      savedPGs: user.savedPGs,
      isBlocked: user.isBlocked,
      avatar: user.avatar,
    },
  });
};

const generateOtp = () => String(Math.floor(100000 + Math.random() * 900000));

const hashOtp = (otp) => crypto.createHash('sha256').update(otp).digest('hex');

const buildOtpEmailHtml = (name, otp, purpose) => {
  const heading = purpose === 'signup' ? 'Verify your account' : 'Verify your login';
  const body =
    purpose === 'signup'
      ? 'Use this OTP to complete your signup.'
      : 'Use this OTP to complete your login.';

  return `
    <div style="font-family: sans-serif; max-width: 600px; margin: auto; padding: 20px;">
      <h2 style="color: #4F46E5;">${heading}</h2>
      <p>Hi ${name || 'there'},</p>
      <p>${body}</p>
      <div style="font-size:30px;font-weight:700;letter-spacing:8px;margin:20px 0;color:#111827;">
        ${otp}
      </div>
      <p>This OTP is valid for <strong>${OTP_TTL_MINUTES} minutes</strong>.</p>
      <p style="color:#666;font-size:13px;">If this wasn't you, you can ignore this email.</p>
    </div>
  `;
};

const setOtpAndSendEmail = async (user, purpose) => {
  const otp = generateOtp();
  user.otpCode = hashOtp(otp);
  user.otpPurpose = purpose;
  user.otpExpiresAt = new Date(Date.now() + OTP_TTL_MINUTES * 60 * 1000);
  await user.save({ validateBeforeSave: false });

  const html = buildOtpEmailHtml(user.name, otp, purpose);
  const subject = purpose === 'signup' ? 'Sikkim PG Finder - Signup OTP' : 'Sikkim PG Finder - Login OTP';

  let emailSent = true;
  try {
    await sendEmail({ to: user.email, subject, html });
  } catch (err) {
    emailSent = false;
    if (process.env.NODE_ENV === 'production') {
      throw err;
    }
    // In development, allow OTP flow to continue even if SMTP is misconfigured.
    console.warn('OTP email send failed in development mode:', err.message);
  }

  return { emailSent };
};

const sendOtpPendingResponse = (res, user, purpose, emailSent = true) => {
  const response = {
    success: true,
    requiresOtp: true,
    purpose,
    email: user.email,
    message: emailSent
      ? `OTP sent to ${user.email}.`
      : 'OTP generated, but email could not be sent. Please check SMTP configuration and try resend.',
  };

  res.status(200).json(response);
};

const isAccountVerified = (user) => user?.isVerified !== false;

// @desc   Register user
// @route  POST /api/auth/signup
export const signup = asyncHandler(async (req, res) => {
  const errors = validationResult(req);
  if (!errors.isEmpty()) {
    return res.status(400).json({ success: false, message: 'Validation failed', errors: errors.array() });
  }

  const { name, email, password } = req.body;

  let user = await User.findOne({ email: email.toLowerCase() }).select('+otpCode +otpPurpose +otpExpiresAt');

  if (user && isAccountVerified(user)) {
    res.status(400);
    throw new Error('Email already registered');
  }

  if (!user) {
    user = await User.create({ name, email, password, isVerified: false });
  } else {
    user.name = name;
    user.password = password;
    user.isVerified = false;
    await user.save();
    user = await User.findById(user._id).select('+otpCode +otpPurpose +otpExpiresAt');
  }

  const { emailSent } = await setOtpAndSendEmail(user, 'signup');
  sendOtpPendingResponse(res, user, 'signup', emailSent);
});

// @desc   Login user
// @route  POST /api/auth/login
export const login = asyncHandler(async (req, res) => {
  const errors = validationResult(req);
  if (!errors.isEmpty()) {
    return res.status(400).json({ success: false, message: 'Validation failed', errors: errors.array() });
  }

  const { email, password } = req.body;

  const user = await User.findOne({ email: email.toLowerCase() }).select('+password +otpCode +otpPurpose +otpExpiresAt');
  if (!user) {
    res.status(401);
    throw new Error('Invalid email or password');
  }

  if (user.isBlocked) {
    res.status(403);
    throw new Error('Your account has been blocked by admin. Please contact support.');
  }

  const isMatch = await user.comparePassword(password);
  if (!isMatch) {
    res.status(401);
    throw new Error('Invalid email or password');
  }

  const { emailSent } = await setOtpAndSendEmail(user, 'login');
  sendOtpPendingResponse(res, user, 'login', emailSent);
});

// @desc   Verify OTP and complete auth
// @route  POST /api/auth/verify-otp
export const verifyOtp = asyncHandler(async (req, res) => {
  const errors = validationResult(req);
  if (!errors.isEmpty()) {
    return res.status(400).json({ success: false, message: 'Validation failed', errors: errors.array() });
  }

  const { email, otp, purpose } = req.body;

  if (!email || !otp || !purpose) {
    res.status(400);
    throw new Error('Email, OTP and purpose are required');
  }

  if (!['signup', 'login'].includes(purpose)) {
    res.status(400);
    throw new Error('Invalid OTP purpose');
  }

  const user = await User.findOne({ email: email.toLowerCase() }).select('+otpCode +otpPurpose +otpExpiresAt');
  if (!user || !user.otpCode || !user.otpExpiresAt || !user.otpPurpose) {
    res.status(400);
    throw new Error('No active OTP. Please request a new one.');
  }

  if (user.otpPurpose !== purpose) {
    res.status(400);
    throw new Error('OTP purpose mismatch. Please request a new OTP.');
  }

  if (user.otpExpiresAt.getTime() < Date.now()) {
    user.otpCode = undefined;
    user.otpExpiresAt = undefined;
    user.otpPurpose = undefined;
    await user.save({ validateBeforeSave: false });
    res.status(400);
    throw new Error('OTP has expired. Please request a new OTP.');
  }

  const hashed = hashOtp(String(otp).trim());
  if (hashed !== user.otpCode) {
    res.status(400);
    throw new Error('Invalid OTP');
  }

  if (purpose === 'signup' && user.isVerified === false) {
    user.isVerified = true;
    await createNotification(
      `New user registered: ${user.name} (${user.email})`,
      'USER_SIGNUP',
      user._id,
      user._id
    );
  }

  user.otpCode = undefined;
  user.otpExpiresAt = undefined;
  user.otpPurpose = undefined;
  await user.save({ validateBeforeSave: false });

  const token = generateToken(user._id, user.role);
  setCookieAndRespond(res, 200, user, token);
});

// @desc   Resend OTP
// @route  POST /api/auth/resend-otp
export const resendOtp = asyncHandler(async (req, res) => {
  const errors = validationResult(req);
  if (!errors.isEmpty()) {
    return res.status(400).json({ success: false, message: 'Validation failed', errors: errors.array() });
  }

  const { email, purpose } = req.body;

  if (!email || !purpose) {
    res.status(400);
    throw new Error('Email and purpose are required');
  }

  if (!['signup', 'login'].includes(purpose)) {
    res.status(400);
    throw new Error('Invalid OTP purpose');
  }

  const user = await User.findOne({ email: email.toLowerCase() }).select('+otpCode +otpPurpose +otpExpiresAt');
  if (!user) {
    res.status(404);
    throw new Error('User not found');
  }

  if (user.isBlocked) {
    res.status(403);
    throw new Error('Your account has been blocked by admin. Please contact support.');
  }

  if (purpose === 'signup' && isAccountVerified(user)) {
    res.status(400);
    throw new Error('Account is already verified. Please login.');
  }

  const { emailSent } = await setOtpAndSendEmail(user, purpose);
  sendOtpPendingResponse(res, user, purpose, emailSent);
});

// @desc   Get current user
// @route  GET /api/auth/me
export const getMe = asyncHandler(async (req, res) => {
  const user = await User.findById(req.user._id).populate('savedPGs', 'title price location status images');
  res.json({ success: true, user });
});

// @desc   Logout
// @route  POST /api/auth/logout
export const logout = asyncHandler(async (req, res) => {
  res.cookie('token', '', { httpOnly: true, expires: new Date(0) });
  res.json({ success: true, message: 'Logged out successfully' });
});

// @desc   Forgot password
// @route  POST /api/auth/forgot-password
export const forgotPassword = asyncHandler(async (req, res) => {
  const { email } = req.body;
  const user = await User.findOne({ email: email?.toLowerCase() });

  // Always return success (security: don't reveal if email exists)
  if (!user) {
    return res.json({ success: true, message: 'If that email exists, a reset link has been sent.' });
  }

  const resetToken = user.getResetPasswordToken();
  await user.save({ validateBeforeSave: false });

  const resetUrl = `${process.env.CLIENT_URL}/reset-password/${resetToken}`;
  const html = `
    <div style="font-family: sans-serif; max-width: 600px; margin: auto; padding: 20px;">
      <h2 style="color: #4F46E5;">Reset Your Password</h2>
      <p>You requested a password reset for your Sikkim PG Finder account.</p>
      <p>Click the button below to reset your password. This link expires in <strong>10 minutes</strong>.</p>
      <a href="${resetUrl}" style="display:inline-block;background:#4F46E5;color:white;padding:12px 24px;border-radius:8px;text-decoration:none;font-weight:600;margin:16px 0;">
        Reset Password
      </a>
      <p style="color:#666;font-size:13px;">If you didn't request this, you can safely ignore this email.</p>
    </div>
  `;

  try {
    await sendEmail({ to: user.email, subject: 'Sikkim PG Finder — Password Reset', html });
    res.json({ success: true, message: 'If that email exists, a reset link has been sent.' });
  } catch (err) {
    user.resetPasswordToken = undefined;
    user.resetPasswordExpire = undefined;
    await user.save({ validateBeforeSave: false });
    res.status(500);
    throw new Error('Email could not be sent. Please try again later.');
  }
});

// @desc   Reset password
// @route  PUT /api/auth/reset-password/:token
export const resetPassword = asyncHandler(async (req, res) => {
  const hashedToken = crypto.createHash('sha256').update(req.params.token).digest('hex');

  const user = await User.findOne({
    resetPasswordToken: hashedToken,
    resetPasswordExpire: { $gt: Date.now() },
  });

  if (!user) {
    res.status(400);
    throw new Error('Invalid or expired reset token');
  }

  user.password = req.body.password;
  user.resetPasswordToken = undefined;
  user.resetPasswordExpire = undefined;
  await user.save();

  const token = generateToken(user._id, user.role);
  setCookieAndRespond(res, 200, user, token);
});

File: backend/app/controllers/auth_controller.py
from datetime import timedelta, timezone

from app.core.exceptions import AppError
from app.core.security import create_access_token, hash_password, verify_password
from app.core.settings import get_settings
from app.models import collections
from app.services.auth_service import generate_otp, hash_value, otp_expiry_datetime, reset_token
from app.services.email_service import send_email
from app.services.notification_service import create_notification
from app.utils.bson_utils import serialize_doc, to_object_id
from app.utils.datetime_utils import utc_now


async def signup_controller(db, payload):
    email = payload.email.lower()
    user = await db[collections.USERS].find_one({"email": email})

    if user and user.get("isVerified", True):
        raise AppError(status_code=400, message="Email already registered")

    now = utc_now()
    base_fields = {
        "name": payload.name,
        "email": email,
        "password": hash_password(payload.password),
        "isVerified": False,
        "updatedAt": now,
    }

    if not user:
        base_fields.update(
            {
                "role": "user",
                "ownerRequestStatus": "none",
                "savedPGs": [],
                "isBlocked": False,
                "createdAt": now,
            }
        )
        insert_res = await db[collections.USERS].insert_one(base_fields)
        user = await db[collections.USERS].find_one({"_id": insert_res.inserted_id})
    else:
        await db[collections.USERS].update_one({"_id": user["_id"]}, {"$set": base_fields})
        user = await db[collections.USERS].find_one({"_id": user["_id"]})

    otp = generate_otp()
    await db[collections.USERS].update_one(
        {"_id": user["_id"]},
        {
            "$set": {
                "otpCode": hash_value(otp),
                "otpPurpose": "signup",
                "otpExpiresAt": otp_expiry_datetime(now),
                "updatedAt": utc_now(),
            }
        },
    )

    html = build_otp_email_html(user.get("name"), otp, "signup")
    email_sent = True
    try:
        await send_email(to=email, subject="Sikkim PG Finder - Signup OTP", html=html)
    except Exception:
        email_sent = False

    return {
        "success": True,
        "requiresOtp": True,
        "purpose": "signup",
        "email": email,
        "message": f"OTP sent to {email}." if email_sent else "OTP generated but email could not be sent.",
    }


async def login_controller(db, payload):
    email = payload.email.lower()
    user = await db[collections.USERS].find_one({"email": email})
    if not user or not verify_password(payload.password, user["password"]):
        raise AppError(status_code=401, message="Invalid email or password")
    if user.get("isBlocked"):
        raise AppError(status_code=403, message="Your account has been blocked by admin. Please contact support.")
    if user.get("isVerified") is False:
        raise AppError(status_code=403, message="Please verify your signup OTP first.")

    token = create_access_token(str(user["_id"]), user["role"])
    safe_user = sanitize_user(user)
    return token, safe_user


async def verify_otp_controller(db, payload):
    email = payload.email.lower()
    user = await db[collections.USERS].find_one({"email": email})
    if not user or not user.get("otpCode") or not user.get("otpExpiresAt"):
        raise AppError(status_code=400, message="No active OTP. Please request a new one.")

    if user.get("otpPurpose") != payload.purpose.value:
        raise AppError(status_code=400, message="OTP purpose mismatch. Please request a new OTP.")

    otp_expires_at = user["otpExpiresAt"]
    if getattr(otp_expires_at, "tzinfo", None) is not None:
        otp_expires_at = otp_expires_at.astimezone(timezone.utc).replace(tzinfo=None)

    if otp_expires_at < utc_now():
        await db[collections.USERS].update_one(
            {"_id": user["_id"]},
            {"$unset": {"otpCode": "", "otpExpiresAt": "", "otpPurpose": ""}},
        )
        raise AppError(status_code=400, message="OTP has expired. Please request a new OTP.")

    if hash_value(payload.otp.strip()) != user["otpCode"]:
        raise AppError(status_code=400, message="Invalid OTP")

    updates = {"updatedAt": utc_now()}
    unset_fields = {"otpCode": "", "otpExpiresAt": "", "otpPurpose": ""}

    if payload.purpose.value == "signup" and user.get("isVerified") is False:
        updates["isVerified"] = True
        await create_notification(
            db,
            message=f"New user registered: {user['name']} ({user['email']})",
            notification_type="USER_SIGNUP",
            user_id=user["_id"],
            related_id=user["_id"],
        )

    await db[collections.USERS].update_one(
        {"_id": user["_id"]},
        {"$set": updates, "$unset": unset_fields},
    )
    updated_user = await db[collections.USERS].find_one({"_id": user["_id"]})

    token = create_access_token(str(updated_user["_id"]), updated_user["role"])
    safe_user = sanitize_user(updated_user)
    return token, safe_user


async def resend_otp_controller(db, payload):
    email = payload.email.lower()
    user = await db[collections.USERS].find_one({"email": email})
    if not user:
        raise AppError(status_code=404, message="User not found")
    if user.get("isBlocked"):
        raise AppError(status_code=403, message="Your account has been blocked by admin. Please contact support.")
    if payload.purpose.value == "signup" and user.get("isVerified", True):
        raise AppError(status_code=400, message="Account is already verified. Please login.")

    otp = generate_otp()
    now = utc_now()
    await db[collections.USERS].update_one(
        {"_id": user["_id"]},
        {
            "$set": {
                "otpCode": hash_value(otp),
                "otpPurpose": payload.purpose.value,
                "otpExpiresAt": otp_expiry_datetime(now),
                "updatedAt": now,
            }
        },
    )

    html = build_otp_email_html(user.get("name"), otp, payload.purpose.value)
    email_sent = True
    try:
        await send_email(
            to=email,
            subject=f"Sikkim PG Finder - {payload.purpose.value.capitalize()} OTP",
            html=html,
        )
    except Exception:
        email_sent = False

    return {
        "success": True,
        "requiresOtp": True,
        "purpose": payload.purpose.value,
        "email": email,
        "message": f"OTP sent to {email}." if email_sent else "OTP generated but email could not be sent.",
    }


async def get_me_controller(db, current_user):
    user = await db[collections.USERS].find_one({"_id": to_object_id(current_user["id"])})
    if not user:
        raise AppError(status_code=404, message="User not found")

    saved_docs = []
    if user.get("savedPGs"):
        cursor = db[collections.PGS].find({"_id": {"$in": user.get("savedPGs", [])}})
        saved_docs = await cursor.to_list(length=500)

    payload = sanitize_user(user)
    payload["savedPGsData"] = [serialize_doc(doc) for doc in saved_docs]
    return {"success": True, "user": payload}


async def forgot_password_controller(db, payload):
    email = payload.email.lower()
    user = await db[collections.USERS].find_one({"email": email})
    if not user:
        return {"success": True, "message": "If that email exists, a reset link has been sent."}

    raw_token = reset_token()
    hashed_token = hash_value(raw_token)
    expire_at = utc_now() + timedelta(minutes=10)
    await db[collections.USERS].update_one(
        {"_id": user["_id"]},
        {"$set": {"resetPasswordToken": hashed_token, "resetPasswordExpire": expire_at, "updatedAt": utc_now()}},
    )

    settings = get_settings()
    reset_url = f"{settings.client_url}/reset-password/{raw_token}"
    html = f"""
    <div style=\"font-family: sans-serif; max-width: 600px; margin: auto; padding: 20px;\">
      <h2 style=\"color: #4F46E5;\">Reset Your Password</h2>
      <p>You requested a password reset for your Sikkim PG Finder account.</p>
      <a href=\"{reset_url}\" style=\"display:inline-block;background:#4F46E5;color:white;padding:12px 24px;border-radius:8px;text-decoration:none;font-weight:600;margin:16px 0;\">Reset Password</a>
      <p style=\"color:#666;font-size:13px;\">If you didn't request this, ignore this email.</p>
    </div>
    """

    try:
        await send_email(to=email, subject="Sikkim PG Finder - Password Reset", html=html)
    except Exception as exc:
        await db[collections.USERS].update_one(
            {"_id": user["_id"]},
            {"$unset": {"resetPasswordToken": "", "resetPasswordExpire": ""}},
        )
        raise AppError(status_code=500, message="Email could not be sent. Please try again later.") from exc

    return {"success": True, "message": "If that email exists, a reset link has been sent."}


async def reset_password_controller(db, token: str, password: str):
    hashed_token = hash_value(token)
    user = await db[collections.USERS].find_one(
        {
            "resetPasswordToken": hashed_token,
            "resetPasswordExpire": {"$gt": utc_now()},
        }
    )
    if not user:
        raise AppError(status_code=400, message="Invalid or expired reset token")

    await db[collections.USERS].update_one(
        {"_id": user["_id"]},
        {
            "$set": {"password": hash_password(password), "updatedAt": utc_now()},
            "$unset": {"resetPasswordToken": "", "resetPasswordExpire": ""},
        },
    )
    updated = await db[collections.USERS].find_one({"_id": user["_id"]})
    token_value = create_access_token(str(updated["_id"]), updated["role"])
    return token_value, sanitize_user(updated)


def build_otp_email_html(name: str | None, otp: str, purpose: str) -> str:
    heading = "Verify your account" if purpose == "signup" else "Verify your login"
    body = "Use this OTP to complete your signup." if purpose == "signup" else "Use this OTP to complete your login."
    ttl = get_settings().otp_expires_minutes
    return f"""
    <div style=\"font-family: sans-serif; max-width: 600px; margin: auto; padding: 20px;\">
      <h2 style=\"color: #4F46E5;\">{heading}</h2>
      <p>Hi {name or 'there'},</p>
      <p>{body}</p>
      <div style=\"font-size:30px;font-weight:700;letter-spacing:8px;margin:20px 0;color:#111827;\">{otp}</div>
      <p>This OTP is valid for <strong>{ttl} minutes</strong>.</p>
    </div>
    """


def sanitize_user(user_doc: dict) -> dict:
    user = serialize_doc(user_doc) or {}
    user.pop("password", None)
    user.pop("otpCode", None)
    user.pop("otpExpiresAt", None)
    user.pop("otpPurpose", None)
    user.pop("resetPasswordToken", None)
    user.pop("resetPasswordExpire", None)
    user.setdefault("savedPGs", [])
    user.setdefault("ownerRequestStatus", "none")
    user.setdefault("isBlocked", False)
    return user
